File: stats.py
"""セッション統計（純粋ロジック、cv2 非依存）。

維持率の分母は「壁時計時間」ではなく「顔が映っていた時間」。席を立つなど顔が
フレームから外れている間はタイマーを止め、目線維持率が汚れないようにする。
"""

from __future__ import annotations

from dataclasses import dataclass, field

from gaze import ContactState

# 逸れ方向コード → 日本語ラベル
_DRIFT_LABELS = {
    "down": "下（画面・手元）",
    "up": "上",
    "left": "左",
    "right": "右",
}


@dataclass
class SessionStats:
    """1 セッション分の目線維持の集計。"""

    total_time: float = 0.0  # 経過時間（全フレーム）
    visible_time: float = 0.0  # 顔が映っていた時間（維持率の分母）
    contact_time: float = 0.0  # カメラ目線だった時間
    current_streak: float = 0.0  # 現在の連続キープ時間
    longest_streak: float = 0.0  # 最長キープ時間
    drift_time: dict[str, float] = field(
        default_factory=lambda: {"down": 0.0, "up": 0.0, "left": 0.0, "right": 0.0}
    )

    def update(
        self,
        dt: float,
        state: ContactState | None,
        drift: str | None = None,
    ) -> None:
        """1 フレーム分を反映する。state=None は顔ロスト（タイマー停止）。"""
        self.total_time += dt
        if state is None:
            self.current_streak = 0.0
            return
        self.visible_time += dt
        if state is ContactState.CONTACT:
            self.contact_time += dt
            self.current_streak += dt
            self.longest_streak = max(self.longest_streak, self.current_streak)
        else:
            self.current_streak = 0.0
            if drift is not None and drift in self.drift_time:
                self.drift_time[drift] += dt

    @property
    def contact_ratio(self) -> float:
        """顔可視時間に対するカメラ目線の割合（0.0–1.0）。"""
        return self.contact_time / self.visible_time if self.visible_time > 0 else 0.0

    def dominant_drift(self) -> str | None:
        """最も長く視線が外れていた向き（無ければ None）。"""
        if not any(self.drift_time.values()):
            return None
        return max(self.drift_time, key=lambda k: self.drift_time[k])

    def reset(self) -> None:
        self.total_time = 0.0
        self.visible_time = 0.0
        self.contact_time = 0.0
        self.current_streak = 0.0
        self.longest_streak = 0.0
        for key in self.drift_time:
            self.drift_time[key] = 0.0

    def summary_lines(self) -> list[str]:
        """終了時にターミナルへ出すサマリー。"""
        lines = [
            "===== セッション結果 =====",
            f"経過時間          : {self.total_time:6.1f} 秒",
            f"顔が映っていた時間: {self.visible_time:6.1f} 秒",
            f"カメラ目線維持率  : {self.contact_ratio * 100:5.1f} %",
            f"最長キープ        : {self.longest_streak:6.1f} 秒",
        ]
        dominant = self.dominant_drift()
        if dominant is not None:
            lines.append(f"逸れがちな向き    : {_DRIFT_LABELS.get(dominant, dominant)}")
        lines.append("==========================")
        return lines

File: web/src/lib/stats.ts
// セッション統計（純粋・DOM 非依存）。stats.py からの移植。
// 維持率の分母は「顔が映っていた時間」（壁時計ではない）。顔ロスト中はタイマーを止める。

import { ContactState } from "./gaze";

// 逸れ方向コード → 日本語ラベル
export const DRIFT_LABELS: Record<string, string> = {
  down: "下（画面・手元）",
  up: "上",
  left: "左",
  right: "右",
};

export class SessionStats {
  totalTime = 0; // 経過時間（全フレーム）
  visibleTime = 0; // 顔が映っていた時間（維持率の分母）
  contactTime = 0; // カメラ目線だった時間
  currentStreak = 0; // 現在の連続キープ時間
  longestStreak = 0; // 最長キープ時間
  driftTime: Record<string, number> = { down: 0, up: 0, left: 0, right: 0 };

  // 1 フレーム分を反映する。state=null は顔ロスト（タイマー停止）。
  update(dt: number, state: ContactState | null, drift: string | null = null): void {
    this.totalTime += dt;
    if (state === null) {
      this.currentStreak = 0;
      return;
    }
    this.visibleTime += dt;
    if (state === ContactState.Contact) {
      this.contactTime += dt;
      this.currentStreak += dt;
      this.longestStreak = Math.max(this.longestStreak, this.currentStreak);
    } else {
      this.currentStreak = 0;
      if (drift !== null && Object.hasOwn(this.driftTime, drift)) {
        this.driftTime[drift] += dt;
      }
    }
  }

  get contactRatio(): number {
    return this.visibleTime > 0 ? this.contactTime / this.visibleTime : 0;
  }

  dominantDrift(): string | null {
    let best: string | null = null;
    let bestValue = 0;
    for (const [key, value] of Object.entries(this.driftTime)) {
      if (value > bestValue) {
        bestValue = value;
        best = key;
      }
    }
    return best;
  }

  reset(): void {
    this.totalTime = 0;
    this.visibleTime = 0;
    this.contactTime = 0;
    this.currentStreak = 0;
    this.longestStreak = 0;
    this.driftTime = { down: 0, up: 0, left: 0, right: 0 };
  }

  summaryLines(): string[] {
    const lines = [
      `経過時間          : ${this.totalTime.toFixed(1)} 秒`,
      `顔が映っていた時間: ${this.visibleTime.toFixed(1)} 秒`,
      `カメラ目線維持率  : ${(this.contactRatio * 100).toFixed(1)} %`,
      `最長キープ        : ${this.longestStreak.toFixed(1)} 秒`,
    ];
    const dominant = this.dominantDrift();
    if (dominant !== null) {
      lines.push(`逸れがちな向き    : ${DRIFT_LABELS[dominant] ?? dominant}`);
    }
    return lines;
  }
}
